cityofaustin/atd-vz-data
# LDM Persons Table Triggers

## Associated issues

https://github.com/cityofaustin/atd-data-tech/issues/13703

## Backfill commands

This command populates the fatalities table from the production read replica.

This won't work until we have a complete set of crashes instead of the subset we have now in the ldm staging database.

```

docker exec -i \
-e PGHOST=db-rr.vision-zero.austinmobility.io \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=$RR_PASSWORD \
-e PGUSER=$RR_USER \
visionzero-postgis \
pg_dump -t public.fatalities --data-only | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on;
```

## Testing

Testing is most easily done in your local database. 
* Get yourself the latest and greatest LDM data via `./vision_zero --ldm replicate-db`
* Step through the migrations as they are applied to see what you have here:

```bash
hasura migrate apply --database-name default  --up 1; hasura migrate --database-name default status;
```

  * `1697646059131_changing-keying-on-ldm-view`
    * Adjust the `public.atd_txdot_person` to only use the actual trio of fields that comprise the primary key of persons records. (`crash_id`, `unit_nbr`, & `prsn_nbr` and not `prsn_type_id` & `prsn_occpnt_pos_id`.)
  * `1697646059132_fatalities_table`
    * Create the fatalities table. We'll back fill this at merge. 
  * `1697646059133_fataility_insert_function`
    * Create the `fatality_insert_function`. Note that this doesn't need a if/else construct because it's going to be used in an on-insert trigger, and those only ever will happen that we care about when we insert into the `cris_facts` schema.
  * `1697646059134_persons_fatality_insert_trigger`
    * Bind the above function to insert on `cris_facts.atd_txdot_persons` table.
  * `1697646059135_rename-existing-trigger`
    * Rename an existing trigger to simply understanding and establishing the order in which they run.
  * `1697646059136_modify-ldm-maint-trigger`
    * Define trigger that is responsible for creating or removing the record in `vz_facts.atd_txdot_persons` when a record is created or removed in the `cris_facts` schema.
  * there is no `1697646059136` migration. I had something here and removed it.
  * `1697646059138_update_fatality_soft_delete_function`
    * Add the soft_delete hander for the fatalities table. This has a branch for when this is called from an edit on the VZ side and on the CRIS side.
  * `1697646059139_update_fatality_soft_delete_trigger`
    * Bind soft_delete handler to both the `vz_facts` and `cris_facts` `atd_txdot_persons` tables.
  * `1697646059140_add_person_audit_log_function`
    * Create the `audit_log` function for storing a backup of a person record when a change is made. This is the old change log tracking system.
  * `1697646059141_add_person_audit_log_trigger`
    * Bind the `audit_log` function to updates to the persons tables in the `vz_facts` and `cris_facts` schemata.

## Helpful commands
Step through migrations, backwards and forwards
```
hasura migrate apply --database-name default  --up 1; hasura migrate --database-name default status
hasura migrate apply --database-name default  --down 1; hasura migrate --database-name default status
```

---
#### Ship list
